dev/tools/native_driver/lib/src/backend/android/adb.dart

```diff
@@ -0,0 +1,318 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io' as io;
+import 'dart:typed_data';
+
+import 'package:meta/meta.dart';
+import 'package:process/process.dart';
+
+/// A minimal wrapper around the `adb` command-line tool.
+@internal
+class Adb {
+  const Adb._(this._prefixArgs, this._proccess);
+
+  /// Creates a new `adb` command runner that uses the `adb` command-line tool.
+  ///
+  /// If [adbPath] is not provided, the `adb` command is assumed to be in the
+  /// system's PATH.
+  ///
+  /// If [target] is not provided, the target is assumed to be the only running
+  /// emulator or connected device.
+  ///
+  /// For testing, [processManager] can be provided to mock the process manager.
+  static Future<Adb> create({
+    String? adbPath,
+    AndroidDeviceTarget? target,
+    @visibleForTesting ProcessManager? processManager,
+  }) async {
+    target ??= const AndroidDeviceTarget.onlyEmulatorOrDevice();
+    final String tool = adbPath ?? 'adb';
+    final Adb adb = Adb._(
+      <String>[tool, ...target._toAdbArgs()],
+      processManager ?? const LocalProcessManager(),
+    );
+    final (bool connected, String? error) = await adb.isDeviceConnected();
+    if (!connected) {
+      throw StateError('No device connected: $error');
+    }
+    return adb;
+  }
+
+  Future<AdbStringResult> _runString(List<String> args) async {
+    final io.ProcessResult result = await _proccess.run(
+      <String>[
+        ..._prefixArgs,
+        ...args,
+      ],
+    );
+    return AdbStringResult(
+      result.stdout as String,
+      exitCode: result.exitCode,
+      stderr: result.stderr as String,
+    );
+  }
+
+  Future<AdbBinaryResult> _runBinary(List<String> args) async {
+    final io.ProcessResult result = await _proccess.run(
+      <String>[
+        ..._prefixArgs,
+        ...args,
+      ],
+      stdoutEncoding: null,
+    );
+    return AdbBinaryResult(
+      result.stdout as Uint8List,
+      exitCode: result.exitCode,
+      stderr: result.stderr as String,
+    );
+  }
+
+  final List<String> _prefixArgs;
+  final ProcessManager _proccess;
+
+  /// Returns whether the device is currently connected.
+  ///
+  /// Returns a tuple of a boolean indicating whether the device is connected
+  /// and an error message if the device is not connected. If the device is
+  /// connected, the error message is `null`.
```

Comment:
Isn't this just an Exception?

Reply:
Sort of? I didn't want to throw, but maybe you're saying this is better done as `ensureDeviceConnected()`?

(I will modify this post-submit at this point)